File: App/app/src/main/java/com/cinema/ticket_booking/ui/search/CinemaDetailFragment.java
package com.cinema.ticket_booking.ui.search;

import android.os.Bundle;
import android.view.*;
import android.widget.*;
import androidx.annotation.*;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.*;
import com.cinema.ticket_booking.R;
import com.cinema.ticket_booking.data.model.response.*;
import com.cinema.ticket_booking.databinding.FragmentCinemaDetailBinding;
import com.cinema.ticket_booking.ui.booking.SelectShowtimeViewModel;
import com.cinema.ticket_booking.util.SnackbarHelper;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.stream.Collectors;
import java.util.Date;
import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class CinemaDetailFragment extends Fragment {

    private FragmentCinemaDetailBinding binding;
    private SelectShowtimeViewModel viewModel;
    private String cinemaId;
    private String cinemaName;
    private final RecyclerView.RecycledViewPool sharedPool = new RecyclerView.RecycledViewPool();

    @Override
    public View onCreateView(@NonNull LayoutInflater i, ViewGroup c, Bundle s) {
        binding = FragmentCinemaDetailBinding.inflate(i, c, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewModel = new ViewModelProvider(this).get(SelectShowtimeViewModel.class);

        if (getArguments() != null) {
            cinemaId = getArguments().getString("cinemaId");
            cinemaName = getArguments().getString("cinemaName");
            binding.tvCinemaName.setText(cinemaName);
        }

        setupDateSelector();
        setupObservers(view);

        // Load showtimes for this cinema specifically
        viewModel.selectCinema(cinemaId);
        viewModel.loadShowtimes(null); // Load all movies for this cinema

        binding.btnBack.setOnClickListener(v -> Navigation.findNavController(view).popBackStack());
    }

    private void setupDateSelector() {
        List<String> dates = viewModel.getNext7Days();
        LinearLayout container = binding.dateContainer;
        container.removeAllViews();
        SimpleDateFormat display = new SimpleDateFormat("EEE\ndd", new Locale("vi", "VN"));

        for (String date : dates) {
            TextView tv = new TextView(requireContext());
            tv.setText(display.format(parseDate(date)));
            tv.setTextAlignment(View.TEXT_ALIGNMENT_CENTER);
            tv.setPadding(32, 20, 32, 20);
            tv.setTextColor(getResources().getColor(android.R.color.white, null));
            tv.setBackgroundResource(R.drawable.bg_date_selector);

            LinearLayout.LayoutParams lp = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT);
            lp.setMargins(12, 0, 12, 0);
            tv.setLayoutParams(lp);

            // Observe selected date to update UI
            viewModel.getSelectedDate().observe(getViewLifecycleOwner(), selected -> {
                boolean isSelected = date.equals(selected);
                tv.setAlpha(isSelected ? 1.0f : 0.5f);
                tv.setScaleX(isSelected ? 1.1f : 1.0f);
                tv.setScaleY(isSelected ? 1.1f : 1.0f);
            });

            tv.setOnClickListener(v -> viewModel.selectDate(date));
            container.addView(tv);
        }
    }

    private void setupObservers(View view) {
        viewModel.getShowtimes().observe(getViewLifecycleOwner(), resource -> {
            switch (resource.status) {
                case LOADING -> binding.progressBar.setVisibility(View.VISIBLE);
                case SUCCESS -> {
                    binding.progressBar.setVisibility(View.GONE);
                    if (resource.data == null || resource.data.isEmpty()) {
                        binding.tvEmpty.setVisibility(View.VISIBLE);
                        binding.rvCinemaMovies.setVisibility(View.GONE);
                    } else {
                        binding.tvEmpty.setVisibility(View.GONE);
                        binding.rvCinemaMovies.setVisibility(View.VISIBLE);

                        // Group showtimes by Movie
                        Map<String, List<ShowtimeResponse>> grouped = resource.data.stream()
                                .collect(Collectors.groupingBy(ShowtimeResponse::getMovieId));

                        List<List<ShowtimeResponse>> moviesList = new ArrayList<>(grouped.values());

                        binding.rvCinemaMovies.setLayoutManager(new LinearLayoutManager(requireContext()));
                        binding.rvCinemaMovies.setAdapter(new CinemaMovieAdapter(moviesList, sharedPool, showtime -> {
                            // Navigation logic
                            SelectShowtimeViewModel.pendingShowtimeId = showtime.getId();
                            SelectShowtimeViewModel.pendingMovieTitle = showtime.getMovieTitle();
                            SelectShowtimeViewModel.pendingShowtimeTime = showtime.getStartTime();
                            SelectShowtimeViewModel.pendingCinemaName = cinemaName;
                            SelectShowtimeViewModel.pendingShowDate = viewModel.getSelectedDate().getValue();

                            Bundle args = new Bundle();
                            args.putString("showtimeId", showtime.getId());
                            Navigation.findNavController(view).navigate(R.id.action_selectShowtime_to_selectSeat, args);
                        }));
                    }
                }
                case ERROR -> {
                    binding.progressBar.setVisibility(View.GONE);
                    SnackbarHelper.showError(binding.getRoot(), resource.message);
                }
            }
        });
    }

    private Date parseDate(String s) {
        try {
            return new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).parse(s);
        } catch (Exception e) {
            return new Date();
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: App/app/src/main/java/com/cinema/ticket_booking/ui/search/CinemaDetailFragment.kt
package com.cinema.ticket_booking.ui.search

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.Navigation
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.cinema.ticket_booking.R
import com.cinema.ticket_booking.data.model.response.ShowtimeResponse
import com.cinema.ticket_booking.databinding.FragmentCinemaDetailBinding
import com.cinema.ticket_booking.ui.booking.SelectShowtimeViewModel
import com.cinema.ticket_booking.util.Resource
import com.cinema.ticket_booking.util.SnackbarHelper
import dagger.hilt.android.AndroidEntryPoint
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

@AndroidEntryPoint
class CinemaDetailFragment : Fragment() {

    private var _binding: FragmentCinemaDetailBinding? = null
    private val binding get() = _binding!!
    private lateinit var viewModel: SelectShowtimeViewModel
    private var cinemaId: String? = null
    private var cinemaName: String? = null
    private val sharedPool = RecyclerView.RecycledViewPool()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentCinemaDetailBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel = ViewModelProvider(this)[SelectShowtimeViewModel::class.java]

        arguments?.let {
            cinemaId = it.getString("cinemaId")
            cinemaName = it.getString("cinemaName")
            binding.tvCinemaName.text = cinemaName
        }

        setupDateSelector()
        setupObservers(view)

        // Load showtimes for this cinema specifically
        viewModel.selectCinema(cinemaId)
        viewModel.loadShowtimes(null) // Load all movies for this cinema

        binding.btnBack.setOnClickListener { Navigation.findNavController(view).popBackStack() }
    }

    private fun setupDateSelector() {
        val dates = viewModel.next7Days
        val container = binding.dateContainer
        container.removeAllViews()
        val display = SimpleDateFormat("EEE\ndd", Locale("vi", "VN"))

        for (date in dates) {
            val tv = TextView(requireContext()).apply {
                text = display.format(parseDate(date))
                textAlignment = View.TEXT_ALIGNMENT_CENTER
                setPadding(32, 20, 32, 20)
                setTextColor(resources.getColor(android.R.color.white, null))
                setBackgroundResource(R.drawable.bg_date_selector)
            }

            val lp = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
            ).apply {
                setMargins(12, 0, 12, 0)
            }
            tv.layoutParams = lp

            // Observe selected date to update UI
            viewModel.selectedDate.observe(viewLifecycleOwner) { selected ->
                val isSelected = date == selected
                tv.alpha = if (isSelected) 1.0f else 0.5f
                tv.scaleX = if (isSelected) 1.1f else 1.0f
                tv.scaleY = if (isSelected) 1.1f else 1.0f
            }

            tv.setOnClickListener { viewModel.selectDate(date) }
            container.addView(tv)
        }
    }

    private fun setupObservers(view: View) {
        viewModel.showtimes.observe(viewLifecycleOwner) { resource ->
            when (resource.status) {
                Resource.Status.LOADING -> binding.progressBar.visibility = View.VISIBLE
                Resource.Status.SUCCESS -> {
                    binding.progressBar.visibility = View.GONE
                    val data = resource.data
                    if (data.isNullOrEmpty()) {
                        binding.tvEmpty.visibility = View.VISIBLE
                        binding.rvCinemaMovies.visibility = View.GONE
                    } else {
                        binding.tvEmpty.visibility = View.GONE
                        binding.rvCinemaMovies.visibility = View.VISIBLE

                        // Group showtimes by Movie
                        val grouped = data.groupBy { it.movieId }
                        val moviesList = ArrayList(grouped.values)

                        binding.rvCinemaMovies.layoutManager = LinearLayoutManager(requireContext())
                        binding.rvCinemaMovies.adapter = CinemaMovieAdapter(moviesList, sharedPool) { showtime ->
                            // Navigation logic
                            SelectShowtimeViewModel.pendingShowtimeId = showtime.id
                            SelectShowtimeViewModel.pendingMovieTitle = showtime.movieTitle
                            SelectShowtimeViewModel.pendingShowtimeTime = showtime.startTime
                            SelectShowtimeViewModel.pendingCinemaName = cinemaName
                            SelectShowtimeViewModel.pendingShowDate = viewModel.selectedDate.value

                            val args = Bundle().apply {
                                putString("showtimeId", showtime.id)
                            }
                            Navigation.findNavController(view).navigate(R.id.action_selectShowtime_to_selectSeat, args)
                        }
                    }
                }
                Resource.Status.ERROR -> {
                    binding.progressBar.visibility = View.GONE
                    SnackbarHelper.showError(binding.root, resource.message ?: "Lỗi tải lịch chiếu")
                }
            }
        }
    }

    private fun parseDate(s: String): Date {
        return try {
            SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).parse(s) ?: Date()
        } catch (e: Exception) {
            Date()
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
